source/slang/slang-preprocessor.cpp

```diff
@@ -2539,6 +2638,25 @@ TokenList preprocessSource(
     preprocessor.fileSystem = desc.fileSystem;
     preprocessor.namePool = desc.namePool;
 
+    // Add builtin macros
+    {
+        auto namePool = desc.namePool;
+
+        const char*const builtinNames[] = { "__FILE__", "__LINE__" };
+        const PreprocessorMacroFlavor builtinFlavors[] = { PreprocessorMacroFlavor::BuiltinFile, PreprocessorMacroFlavor::BuiltinLine };
+
+        for (Index i = 0; i < SLANG_COUNT_OF(builtinNames); i++)
+        {
+            auto name = namePool->getName(builtinNames[i]);
+
+            PreprocessorMacro* macro = CreateMacro(&preprocessor);
+            macro->flavor = builtinFlavors[i];
+            macro->nameAndLoc = NameLoc(name);
+
+            preprocessor.globalEnv.macros[name] = macro;
```

Comment:
Note: We could in theory have a "super-global" macro environment that sits above all user-defined macros. That might allow us to avoid having `#undef` remove these built-in macros.